File: app/src/main/java/com/itsvks/layouteditor/adapters/ColorResourceAdapter.java
package com.itsvks.layouteditor.adapters;

import android.graphics.Color;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.RecyclerView;
import com.blankj.utilcode.util.ClipboardUtils;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.itsvks.layouteditor.ProjectFile;
import com.itsvks.layouteditor.R;
import com.itsvks.layouteditor.adapters.models.ValuesItem;
import com.itsvks.layouteditor.databinding.LayoutValuesItemBinding;
import com.itsvks.layouteditor.databinding.LayoutValuesItemDialogBinding;
import com.itsvks.layouteditor.editor.dialogs.ColorDialog;
import com.itsvks.layouteditor.utils.BitmapUtil;
import com.itsvks.layouteditor.utils.FileUtil;
import com.itsvks.layouteditor.utils.NameErrorChecker;
import com.itsvks.layouteditor.utils.SBUtils;
import java.util.ArrayList;
import java.util.List;

public class ColorResourceAdapter extends RecyclerView.Adapter<ColorResourceAdapter.VH> {

  private List<ValuesItem> colorList = new ArrayList<>();
  private ProjectFile project;

  public ColorResourceAdapter(ProjectFile project, List<ValuesItem> colorList) {
    this.project = project;
    this.colorList = colorList;
  }

  public class VH extends RecyclerView.ViewHolder {
    LayoutValuesItemBinding binding;
    TextView colorName;
    TextView colorValue;

    public VH(@NonNull LayoutValuesItemBinding binding) {
      super(binding.getRoot());
      this.binding = binding;

      colorName = binding.name;
      colorValue = binding.value;
    }
  }

  @Override
  public VH onCreateViewHolder(ViewGroup parent, int viewType) {
    return new VH(
        LayoutValuesItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
  }

  @Override
  public void onBindViewHolder(VH holder, int position) {
    holder.colorName.setText(colorList.get(position).name);
    holder.colorValue.setText(colorList.get(position).value);
    holder
        .binding
        .getRoot()
        .setAnimation(
            AnimationUtils.loadAnimation(
                holder.itemView.getContext(), R.anim.project_list_animation));
    holder
        .binding
        .getRoot()
        .setBackgroundColor(Color.parseColor(colorList.get(position).value));
    BitmapUtil.setTextColorAccordingToBackground(holder.binding.getRoot(), holder.colorName);
    BitmapUtil.setImageTintAccordingToBackground(holder.binding.menu, holder.binding.getRoot());
    if (BitmapUtil.getLuminance(holder.binding.getRoot()) >= 0.5) {
      holder.colorValue.setTextColor(Color.parseColor("#FF313131"));
      holder.binding.versions.setTextColor(Color.parseColor("#FF313131"));
    } else {
      holder.colorValue.setTextColor(Color.parseColor("#FFD9D9D9"));
      holder.binding.versions.setTextColor(Color.parseColor("#FFD9D9D9"));
    }
    holder.binding.menu.setOnClickListener(v -> showOptions(v, position));
    holder
        .binding
        .getRoot()
        .setOnClickListener(
            v -> {
              editColor(v, position);
            });
  }

  @Override
  public int getItemCount() {
    return colorList.size();
  }

  public void generateColorsXml() {
    String colorsPath = project.getColorsPath();

    StringBuilder sb = new StringBuilder();
    sb.append("<resources>\n");
    for (ValuesItem colorItem : colorList) {
      // Generate color item code
      sb.append("\t<color name=\"")
          .append(colorItem.name)
          .append("\">")
          .append(colorItem.value)
          .append("</color>\n");
    }
    sb.append("</resources>");

    FileUtil.writeFile(colorsPath, sb.toString().trim());
  }

  private void showOptions(View v, int position) {
    final PopupMenu popupMenu = new PopupMenu(v.getContext(), v);
    popupMenu.inflate(R.menu.menu_values);
    popupMenu.setOnMenuItemClickListener(
        new PopupMenu.OnMenuItemClickListener() {

          @Override
          public boolean onMenuItemClick(MenuItem item) {

            var id = item.getItemId();
            switch (id) {
              case R.id.menu_copy_name:
                ClipboardUtils.copyText(colorList.get(position).name);
                SBUtils.make(
                        v,
                        v.getContext().getString(R.string.copied)
                            + " "
                            + colorList.get(position).name)
                    .setSlideAnimation()
                    .showAsSuccess();
                return true;
              case R.id.menu_delete:
                new MaterialAlertDialogBuilder(v.getContext())
                    .setTitle("Remove Color")
                    .setMessage("Do you want to remove " + colorList.get(position).name + "?")
                    .setNegativeButton(R.string.no, null)
                    .setPositiveButton(
                        R.string.yes,
                        (d, w) -> {
                          var name = colorList.get(position).name;
                          if (name.equals("default_color")) {
                            SBUtils.make(
                                    v,
                                    v.getContext()
                                        .getString(
                                            R.string.msg_cannot_delete_default, "color"))
                                .setFadeAnimation()
                                .setType(SBUtils.Type.INFO)
                                .show();
                          } else {
                            colorList.remove(position);
                            notifyDataSetChanged();
                            generateColorsXml();
                          }
                        })
                    .show();
                return true;
            }
            return false;
          }
        });

    popupMenu.show();
  }

  private void editColor(View v, int pos) {
    MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(v.getContext());
    builder.setTitle("Edit Color");

    LayoutValuesItemDialogBinding bind =
        LayoutValuesItemDialogBinding.inflate(builder.create().getLayoutInflater());
    TextInputLayout ilName = bind.textInputLayoutName;
    TextInputLayout ilValue = bind.textInputLayoutValue;
    TextInputEditText etName = bind.textinputName;
    TextInputEditText etValue = bind.textinputValue;

    etName.setText(colorList.get(pos).name);
    etValue.setText(colorList.get(pos).value);
    etValue.setFocusable(false);
    builder.setView(bind.getRoot());
    etValue.setOnClickListener(
        (view) -> {
          // ColorPicker
          ColorDialog dialog = new ColorDialog(v.getContext(), etValue.getText().toString());
          dialog.setOnSaveValueListener(
              value -> {
                etValue.setText(value);
              });
          dialog.show();
        });

    builder.setPositiveButton(
        R.string.okay,
        (dlg, i) -> {
          if (colorList.get(pos).name.equals("default_color") && !etName.getText().toString().equals("default_color")) {
            SBUtils.make(v, v.getContext().getString(R.string.msg_cannot_rename_default, "color"))
                .setFadeAnimation()
                .setType(SBUtils.Type.INFO)
                .show();
          } else {
            // Update position
            colorList.get(pos).name = etName.getText().toString();
          }
          // Update position
          colorList.get(pos).value = etValue.getText().toString();
          notifyDataSetChanged();
          // Generate code from all colors in list
          generateColorsXml();
        });
    builder.setNegativeButton(R.string.cancel, null);

    AlertDialog dialog = builder.create();
    dialog.show();

    etName.addTextChangedListener(
        new TextWatcher() {

          @Override
          public void beforeTextChanged(CharSequence p1, int p2, int p3, int p4) {}

          @Override
          public void onTextChanged(CharSequence p1, int p2, int p3, int p4) {}

          @Override
          public void afterTextChanged(Editable p1) {
            NameErrorChecker.checkForValues(
                etName.getText().toString(), ilName, dialog, colorList, pos);
          }
        });
    NameErrorChecker.checkForValues(etName.getText().toString(), ilName, dialog, colorList, pos);
  }
}

File: app/src/main/java/com/itsvks/layouteditor/adapters/ColorResourceAdapter.kt
package com.itsvks.layouteditor.adapters

import android.annotation.SuppressLint
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.appcompat.widget.PopupMenu
import androidx.appcompat.widget.TooltipCompat
import androidx.core.graphics.ColorUtils
import androidx.recyclerview.widget.RecyclerView
import com.blankj.utilcode.util.ClipboardUtils
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.itsvks.layouteditor.ProjectFile
import com.itsvks.layouteditor.R
import com.itsvks.layouteditor.adapters.models.ValuesItem
import com.itsvks.layouteditor.databinding.LayoutColorItemBinding
import com.itsvks.layouteditor.databinding.LayoutValuesItemDialogBinding
import com.itsvks.layouteditor.tools.ColorPickerDialogFlag
import com.itsvks.layouteditor.utils.FileUtil
import com.itsvks.layouteditor.utils.NameErrorChecker
import com.itsvks.layouteditor.utils.SBUtils
import com.skydoves.colorpickerview.ColorPickerDialog
import com.skydoves.colorpickerview.listeners.ColorEnvelopeListener

class ColorResourceAdapter(
  private val project: ProjectFile,
  private val colorList: MutableList<ValuesItem>
) : RecyclerView.Adapter<ColorResourceAdapter.VH>() {

  class VH(var binding: LayoutColorItemBinding) : RecyclerView.ViewHolder(
    binding.root
  ) {
    val colorName: TextView = binding.colorName
    val colorValue: TextView = binding.colorValue
    val colorPreview = binding.colorPreview
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
    return VH(
      LayoutColorItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
    )
  }

  override fun onBindViewHolder(holder: VH, position: Int) {
    holder.itemView.animation = AnimationUtils.loadAnimation(
      holder.itemView.context, R.anim.project_list_animation
    )

    holder.colorName.text = colorList[position].name
    holder.colorValue.text = colorList[position].value

    holder.colorPreview.setImageDrawable(
      drawCircle(
        Color.parseColor(colorList[position].value)
      )
    )

    TooltipCompat.setTooltipText(holder.itemView, colorList[position].name)
    TooltipCompat.setTooltipText(holder.binding.menu, "Options")

    holder.binding.menu.setOnClickListener { showOptions(it, position) }
    holder.itemView.setOnClickListener { editColor(it, position) }
  }

  override fun getItemCount(): Int {
    return colorList.size
  }

  fun generateColorsXml() {
    val colorsPath = project.colorsPath
    val sb = StringBuilder()
    sb.append("<resources>\n")
    for (colorItem in colorList) {
      // Generate color item code
      sb.append("\t<color name=\"")
        .append(colorItem.name)
        .append("\">")
        .append(colorItem.value)
        .append("</color>\n")
    }
    sb.append("</resources>")
    FileUtil.writeFile(colorsPath, sb.toString().trim { it <= ' ' })
  }

  private fun showOptions(v: View, position: Int) {
    val popupMenu = PopupMenu(v.context, v)
    popupMenu.inflate(R.menu.menu_values)
    popupMenu.setOnMenuItemClickListener {
      val id = it.itemId
      when (id) {
        R.id.menu_copy_name -> {
          ClipboardUtils.copyText(colorList[position].name)
          SBUtils.make(
            v, "${v.context.getString(R.string.copied)} ${colorList[position].name}"
          ).setSlideAnimation().showAsSuccess()
          true
        }

        R.id.menu_delete -> {
          MaterialAlertDialogBuilder(v.context)
            .setTitle("Remove Color")
            .setMessage("Do you want to remove ${colorList[position].name}?")
            .setNegativeButton(R.string.no, null)
            .setPositiveButton(
              R.string.yes
            ) { _, _ ->
              val name = colorList[position].name
              if (name == "default_color") {
                SBUtils.make(
                  v,
                  v.context
                    .getString(
                      R.string.msg_cannot_delete_default, "color"
                    )
                ).setFadeAnimation().setType(SBUtils.Type.INFO).show()
              } else {
                colorList.removeAt(position)
                notifyItemRemoved(position)
                generateColorsXml()
              }
            }
            .show()
          true
        }

        else -> false
      }
    }
    popupMenu.show()
  }

  @SuppressLint("SetTextI18n")
  private fun editColor(v: View, pos: Int) {
    val builder = MaterialAlertDialogBuilder(v.context)
    builder.setTitle("Edit Color")
    val bind = LayoutValuesItemDialogBinding.inflate(builder.create().layoutInflater)
    val ilName = bind.textInputLayoutName
    val etName = bind.textinputName
    val etValue = bind.textinputValue
    etName.setText(colorList[pos].name)
    etValue.setText(colorList[pos].value)
    etValue.isFocusable = false
    builder.setView(bind.getRoot())
    etValue.setOnClickListener {
      val dialog: ColorPickerDialog.Builder = ColorPickerDialog.Builder(v.context)
        .setTitle("Choose Color")
        .setPositiveButton(v.context.getString(R.string.confirm),
          ColorEnvelopeListener { envelope, _ ->
            etValue.setText("#${envelope.hexCode}")
          }
        )
        .setNegativeButton(v.context.getString(R.string.cancel)) { d, _ -> d.dismiss() }
        .attachAlphaSlideBar(true)
        .attachBrightnessSlideBar(true)
        .setBottomSpace(12)

      val colorView = dialog.colorPickerView
      colorView.setFlagView(ColorPickerDialogFlag(v.context))
      colorView.setInitialColor(Color.parseColor(colorList[pos].value))
      dialog.show()
    }
    builder.setPositiveButton(
      R.string.okay
    ) { _, _ ->
      if (colorList[pos].name == "default_color" && etName.getText()
          .toString() != "default_color"
      ) {
        SBUtils.make(v, v.context.getString(R.string.msg_cannot_rename_default, "color"))
          .setFadeAnimation()
          .setType(SBUtils.Type.INFO)
          .show()
      } else {
        // Update position
        colorList[pos].name = etName.getText().toString()
      }
      // Update position
      colorList[pos].value = etValue.getText().toString()
      notifyItemChanged(pos)
      // Generate code from all colors in list
      generateColorsXml()
    }
    builder.setNegativeButton(R.string.cancel, null)
    val dialog = builder.create()
    dialog.show()
    etName.addTextChangedListener(
      object : TextWatcher {
        override fun beforeTextChanged(p1: CharSequence, p2: Int, p3: Int, p4: Int) {}
        override fun onTextChanged(p1: CharSequence, p2: Int, p3: Int, p4: Int) {}
        override fun afterTextChanged(p1: Editable) {
          NameErrorChecker.checkForValues(
            etName.getText().toString(), ilName, dialog, colorList, pos
          )
        }
      })
    NameErrorChecker.checkForValues(etName.getText().toString(), ilName, dialog, colorList, pos)
  }

  private fun drawCircle(@ColorInt backgroundColor: Int): Drawable {
    return GradientDrawable().apply {
      shape = GradientDrawable.OVAL
      cornerRadii = floatArrayOf(0f, 0f, 0f, 0f, 0f, 0f, 0f, 0f)
      setColor(backgroundColor)
      setStroke(
        2, if (ColorUtils.calculateLuminance(backgroundColor) >= 0.5) {
          Color.parseColor("#FF313131")
        } else {
          Color.parseColor("#FFD9D9D9")
        }
      )
    }
  }
}
